scenario/context.py

```diff
@@ -41,6 +41,20 @@ class ActionOutput:
     failure: Optional[str] = None
     """If the action is not a success: the message the charm set when failing the action."""
 
+    # NOTE: This can be replaced with dataclass(kw_only=True) in Python 3.10+
+    def __init__(
+        self,
+        *,
+        state: "State",
+        logs: List[str],
+        results: Optional[Dict[str, Any]],
+        failure: Optional[str],
```

Comment:
Yes, fixed.

This isn't caught because I didn't add new tests (maybe I should? But only once we have decided for sure to do this!). The existing code always provides all arguments.

This class is unlike the others in that it seems unlikely that a charmer would ever instantiate it themselves - it's always Scenario creating the object in a `run_action` and returning it to the charm. So perhaps none of the arguments should be optional/have defaults? It's not useful for the internal code, and we don't expect external use, so it seems unnecessary (this would be an additional backwards incompatible change).